src/coreclr/jit/lower.cpp

```diff
@@ -5551,6 +5551,17 @@ GenTree* Lowering::LowerDelegateInvoke(GenTreeCall* call)
     BlockRange().Remove(thisArgNode);
     BlockRange().InsertBefore(call, newThisAddr, newThis, thisArgNode);
 
+#if !defined(TARGET_XARCH)
+    // Keep delegate alive if the target can't do an indirect call with an immediate operand
+    // and only in fully interruptible code.
+    if (comp->GetInterruptible())
+    {
+        GenTree* baseClone     = comp->gtCloneExpr(base);
+        GenTree* keepBaseAlive = comp->gtNewKeepAliveNode(baseClone);
+        BlockRange().InsertBefore(call, baseClone, keepBaseAlive);
```

Comment:
This `InsertBefore` only works due to implementation details of the backend (lazy liveness updates for GC information about registers), which is quite subtle. But of course the fully ideal solution requires adding a new operand to `GenTreeCall` and special casing it throughout the backend as an invisible use, which is probably a bit overkill.

I wonder if we in the uncontained case wouldn't be better off just by inserting `GT_START_NONGC` and creating a nogc region on the single call instruction. Given that we already have a call and hence a GC safe point this does not seem like it would be very harmful; in a less conservative JIT this basic block might have simply always been partially interruptible for that reason.

Reply:
hm.. interesting idea! applied

Reply:
The new diff for the snippet in the issue: https://www.diffchecker.com/CiuxVSBL/

Reply:
I think you will also need to add a `STOP_NOGC` to turn it off again. Looks like we don't have that today.

Reply:
> I think you will also need to add a `STOP_NOGC` to turn it off again. Looks like we don't have that today.

I guess it's not from a correctness stand point, just to reduce the nongc area if we have a lot more statements in the current block?

Reply:
I guess this one is tricky - we might end whatever was expected to be nongc even before we emitted the start_nogc

Reply:
It's counted, so it doesn't turn off nogc until the count gets to zero. But regardless I wouldn't expect there to ever be another active nogc region in this case

Reply:
> I guess it's not from a correctness stand point, just to reduce the nongc area if we have a lot more statements in the current block?

I'm not really sure how the VM behaves if the return address is in a nogc region – but in any case it's unnecessary. I think we only need it to be active from the load of the target address (which is the last use) to the call – should be just one instruction